### PR TITLE
[Model] Add VoxCPM2 startup LoRA adapter support

### PR DESCRIPTION
**Pull request by Bezdarnost**
## Purpose

Allow single-adapter VoxCPM2 deployments to load a native training export directly
at startup, without creating a separate merged model checkpoint.

Set `voxcpm2_runtime_config.startup_lora_path` in the stage's `hf_overrides`.
The loader validates the native `lora_config.json` and `lora_weights.safetensors`,
then merges the configured base LM, residual LM, LocDiT, and projection targets
after base weight loading and before inference optimizations. Base weights retain
their dtype and Parameter identity. Invalid checkpoint keys, ranks, shapes,
dtypes, or non-finite adapter tensors fail loading. Dummy base weights and
subsequent weight reloads are rejected.

Related to #6004. This PR supports one adapter for all requests in a server.
It does not implement the per-request multi-adapter deployment described in that
issue; changing the adapter requires a restart.

## Test Plan

Requires the project's Python 3.12 development environment. GPU tests also require
CUDA, `voxcpm==2.0.3`, the VoxCPM2 base weights, and ffmpeg for the Whisper judge.
The new end-to-end fixture generates small nonzero native adapter weights for all
four target groups; no external fine-tuned checkpoint is required for CI.

```bash
python -m pytest -q -o addopts= tests/model_executor/models/voxcpm2 tests/entrypoints/openai_api/test_serving_speech_stream.py
python -m pytest -s -v -o addopts= tests/e2e/online_serving/test_voxcpm2_tts_lora.py -m 'slow and cuda and tts' --run-level full_model
pre-commit run
```

The GPU test follows the existing VoxCPM2 `slow`/`tts`/L4 classification and is
collected by the weekly TTS L4 sweep. The new CPU tests are collected by the
existing `core_model and cpu` jobs. No Buildkite YAML changes are needed.

For the local RTX 5080 run, `MODEL_PREFIX=/tmp/voxcpm2-models` selected cached base
weights and `VLLM_USE_FLASHINFER_SAMPLER=0` disabled an optional sampler whose
architecture detection failed in this environment. Neither is required by the
feature. The fixture config uses four sequences, a 1 GiB KV cache, BF16, and
unified decode graphs with a maximum batch size of four.

**vLLM Version:** 0.28.0

**vLLM-Omni Commit:** 44d3ae100afa8411770b8b4d9442318bc5e897b3 + this change

**Local environment:** RTX 5080 16 GB; driver 580.159.04; Python 3.12.11;
PyTorch 2.13.0 / CUDA 13.0; transformers 5.14.1; voxcpm 2.0.3.

Base weights: `openbmb/VoxCPM2` revision
`32279effe8c19989596f05d353d1447f51d9e915`. The additional trained-adapter check
used `Saksham1918/voxcpm2-punjabi-lora` revision
`17d6ae532173f468d7223770b2007f4fe1a5a133`.

## Test Result

- Unit and speech-stream suite: 80 passed, including 16 new adapter tests.
- GPU regression suite: 2 passed in 210.71 seconds, including fixture startup
  and cleanup. All five generated outputs scored 1.000 in the repository's
  Whisper/text-similarity check (four concurrent requests and one streaming
  request). The generated adapter merged into 196 linear layers.
- All applicable changed-file pre-commit checks passed, including mypy,
  markdownlint, SPDX, and test markers.
- A public trained adapter, `Saksham1918/voxcpm2-punjabi-lora`, additionally
  loaded into 192 layers and generated non-silent, finite 48 kHz audio.
- Manual HTTP checks covered WAV, raw PCM streaming, SSE delta/done events,
  four simultaneous requests with distinct text, and cancellation recovery.
  Repeated with segmented and unified graphs; unified graph captures at batch
  sizes one and four succeeded. Whisper-base.en recovered the requested text
  from all six samples in each graph-mode check, ignoring punctuation and case.
- Real-weight layer comparison against upstream `LoRALinear` in BF16:
  maximum relative L2 error 0.003661 across 192 layers. In a same-seed native-model
  comparison, runtime LoRA generated 3.68 seconds and merged LoRA 3.52 seconds;
  both transcribed to the requested sentence. Merging changes rounding, so this
  is not a claim of bitwise or general perceptual equivalence.

Full repository GPU CI, long-duration load tests, and multi-GPU execution were
not run locally. No performance improvement is claimed.

### CODE REVIEW

**Comment by Bezdarnost** [comment]
> This PR appears to belong to: **docs/design/module/model_integration.md**.
> 
> Module owners: @tzhouam @gcanlin
> 
> @Bezdarnost, please review your own changes and leave a short self-review comment describing what you checked. PRs without author self-review may not be assigned a reviewer.
> 
> Please take a look when you have a chance. If you would like an automated review, mention @vllm-omni-review-bot in a comment.

Self-review for 58d733e:

- Checked adapter validation, target selection, and the merge formula. Invalid checkpoint inputs are rejected before modifying weights.
- Verified that merging happens after base weight loading and before inference optimizations, preserving Parameter identity and dtype. Dummy loading and repeated weight loading are rejected.
- Local validation passed: 80 unit/streaming tests, 2 full-model GPU tests on RTX 5080, and all applicable pre-commit hooks. The GPU tests cover concurrent and streaming speech requests.
- This PR adds single-adapter loading at startup. Changing adapters requires a server restart. It is a step toward #6004, but does not resolve that issue.

AI tools assisted with implementation and validation. Upstream CI and maintainer review remain pending.

@vllm-omni-review-bot please review.

**Review by linyueqian — APPROVED** [review]
Approving. Merging a LoRA into base weights in place is the kind of operation where a half-finished failure is worse than no failure at all, and this gets the ordering right: every check runs to completion before `torch.no_grad()` is entered, so a malformed adapter raises with the model's weights untouched rather than partway through a layer sweep.

The validation is thorough in the places that matter rather than uniformly. Rank must be a positive int and alpha finite, with `isinstance(alpha, bool)` excluded explicitly, which is the trap Python's bool-is-int inheritance sets for exactly this kind of config parsing. Target names are checked to exist and the merge raises when any configured name matches nothing, so a typo in `target_modules_lm` fails loudly instead of silently merging a subset. The checkpoint key set is compared for exact equality in both directions and reports missing and unexpected keys separately, so a mismatched adapter tells you which way it is wrong. Per-tensor checks cover shape against `(rank, in_features)` and `(out_features, rank)`, floating-point dtype, and finiteness, and the target modules are checked to be materialized rather than on meta.

Two smaller choices worth acknowledging. Using `safetensors.load_file` rather than `torch.load` means an adapter path cannot execute code, which matters because this is a user-supplied path. And merging through `torch.addmm(module.weight.float(), b, a, alpha=alpha / rank)` followed by `copy_` computes in fp32 while preserving the existing Parameter object and its dtype, rather than rebinding the attribute, so anything already holding a reference to that weight stays valid.

CI: `buildkite/vllm-omni` green at this head. Informational lanes carry their standing reds, covered by `ci_approval_override_head`.

Validation: static review of the six-file author diff from the GitHub file list, the validate-then-mutate ordering, the LoRA shape and scaling convention, and the startup call site. Fork head, so the branch was not executed. Not an e2e candidate: a wrong merge would be caught by the shape and finiteness checks rather than by listening to output, and the failure this guards against is a malformed adapter rather than degraded audio.

**Comment by MinhaoLi0318** [comment]
@Bezdarnost Thanks! I’ll pick this up. I’ll first go through #7108 and understand the current startup-time LoRA path, then propose a small first step toward per-request multi-LoRA support. I’ll post the scoped plan here before starting the implementation.
